File: src/tgvoip/tgvoip.py
# PytgVoIP - Telegram VoIP Library for Python
# Copyright (C) 2020 bakatrouble <https://github.com/bakatrouble>
#
# This file is part of PytgVoIP.
#
# PytgVoIP is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PytgVoIP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with PytgVoIP.  If not, see <http://www.gnu.org/licenses/>.

import json
import os
import sys
from datetime import datetime
from enum import Enum
from typing import Union, List

from _tgvoip import (
    NetType as _NetType,
    DataSaving as _DataSaving,
    CallState as _CallState,
    CallError as _CallError,
    Stats,
    Endpoint,
    VoIPController as _VoIPController,
    VoIPServerConfig as _VoIPServerConfig
)

from tgvoip.utils import get_real_elapsed_time


# docstring magic ahead


class NetType(Enum):
    """
    An enumeration of network types

    Members:
        * UNKNOWN = 0
        * GPRS = 1
        * EDGE = 2
        * NET_3G = 3
        * HSPA = 4
        * LTE = 5
        * WIFI = 6
        * ETHERNET = 7
        * OTHER_HIGH_SPEED = 8
        * OTHER_LOW_SPEED = 9
        * DIALUP = 10
        * OTHER_MOBILE = 11
    """

    UNKNOWN = _NetType.UNKNOWN
    GPRS = _NetType.GPRS
    EDGE = _NetType.EDGE
    NET_3G = _NetType.NET_3G
    HSPA = _NetType.HSPA
    LTE = _NetType.LTE
    WIFI = _NetType.WIFI
    ETHERNET = _NetType.ETHERNET
    OTHER_HIGH_SPEED = _NetType.OTHER_HIGH_SPEED
    OTHER_LOW_SPEED = _NetType.OTHER_LOW_SPEED
    DIALUP = _NetType.DIALUP
    OTHER_MOBILE = _NetType.OTHER_MOBILE


class CallState(Enum):
    """
    An enumeration of call states

    Members:
        * WAIT_INIT = 1
        * WAIT_INIT_ACK = 2
        * ESTABLISHED = 3
        * FAILED = 4
        * RECONNECTING = 5
        * HANGING_UP = 10
        * ENDED = 11
        * EXCHANGING_KEYS = 12
        * WAITING = 13
        * REQUESTING = 14
        * WAITING_INCOMING = 15
        * RINGING = 16
        * BUSY = 17
    """

    WAIT_INIT = _CallState.WAIT_INIT
    WAIT_INIT_ACK = _CallState.WAIT_INIT_ACK
    ESTABLISHED = _CallState.ESTABLISHED
    FAILED = _CallState.FAILED
    RECONNECTING = _CallState.RECONNECTING
    HANGING_UP = 10
    ENDED = 11
    EXCHANGING_KEYS = 12
    WAITING = 13
    REQUESTING = 14
    WAITING_INCOMING = 15
    RINGING = 16
    BUSY = 17


class DataSaving(Enum):
    """
    An enumeration of data saving modes

    Members:
        * NEVER = 0
        * MOBILE = 1
        * ALWAYS = 2
    """
    NEVER = _DataSaving.NEVER
    MOBILE = _DataSaving.MOBILE
    ALWAYS = _DataSaving.ALWAYS


class CallError(Enum):
    """
    An enumeration of call errors

    Members:
        * UNKNOWN = 0
        * INCOMPATIBLE = 1
        * TIMEOUT = 2
        * AUDIO_IO = 3
        * PROXY = 4
    """
    UNKNOWN = _CallError.UNKNOWN
    INCOMPATIBLE = _CallError.INCOMPATIBLE
    TIMEOUT = _CallError.TIMEOUT
    AUDIO_IO = _CallError.AUDIO_IO
    PROXY = _CallError.PROXY


class VoIPController(_VoIPController):
    """
    A wrapper around C++ wrapper for libtgvoip ``VoIPController``

    Args:
        persistent_state_file (``str``, *optional*): ?, empty to not use
        debug (``bool``, *optional*): Modifies logging behavior
        logs_dir (``str``, *optional*): Logs directory

    Class attributes:
        LIBTGVOIP_VERSION
            Used ``libtgvoip`` version

        CONNECTION_MAX_LAYER
            Maximum layer supported by used ``libtgvoip`` version

    Attributes:
        persistent_state_file:
            Value set in the constructor

        call_state_changed_handlers:
            ``list`` of call state change callbacks, callbacks receive a :class:`CallState` object as argument

        signal_bars_changed_handlers
            ``list`` of signal bars count change callbacks, callbacks receive an ``int`` object as argument
    """

    def __init__(self, persistent_state_file: str = '', debug=False, logs_dir='logs'):
        super().__init__(persistent_state_file)  # _VoIPController.__init__(self, persistent_state_file)
        self.debug = debug
        self.logs_dir = logs_dir
        self.start_time = 0
        self.send_audio_frame_callback = lambda length: b''
        self.recv_audio_frame_callback = lambda frame: ...
        self.call_state_changed_handlers = []
        self.signal_bars_changed_handlers = []
        self._init()

    @property
    def call_duration(self) -> int:
        """
        Current call duration in seconds as ``int`` if call was started, otherwise 0
        """
        return int(get_real_elapsed_time() - self.start_time) if self.start_time else 0

    def start(self):
        """
        Start the controller
        """
        super().start()

    def connect(self):
        """
        Start the call
        """
        super().connect()
    
    def stop(self):
        """
        Stop the controller
        """
        super().stop()

    def set_proxy(self, address: str, port: int = 1080, username: str = '', password: str = ''):
        """
        Set SOCKS5 proxy config

        Args:
            address (``str``): Proxy hostname or IP address
            port (``int``, *optional*): Proxy port
            username (``int``, *optional*): Proxy username
            password (``int``, *optional*): Proxy password

        Raises:
            :class:`ValueError` if :attr:`address` is empty
        """
        if not address:
            raise ValueError('address can\'t be empty')
        super().set_proxy(address, port, username, password)

    def set_encryption_key(self, key: bytes, is_outgoing: bool):
        """
        Set call auth key

        Args:
            key (``bytes``): Auth key, must be exactly 256 bytes
            is_outgoing (``bool``): Is call outgoing

        Raises:
            :class:`ValueError` if provided auth key has wrong length
        """
        if len(key) != 256:
            raise ValueError('key length must be exactly 256 bytes but is {}'.format(len(key)))
        super().set_encryption_key(key, is_outgoing)

    def set_remote_endpoints(self, endpoints: List[Endpoint], allow_p2p: bool, tcp: bool, connection_max_layer: int):
        """
        Set remote endpoints received in call object from Telegram.

        Usually it's ``[call.connection] + call.alternative_connections``.

        You must build :class:`Endpoint` objects from MTProto :class:`phoneConnection` objects and pass them in list.

        Args:
            endpoints (``list`` of :class:`Endpoint`): List of endpoints
            allow_p2p (``bool``): Is p2p connection allowed, usually `call.p2p_allowed` value is used
            tcp (``bool``): Connect via TCP, not recommended
            connection_max_layer (``int``): Use a value provided by :attr:`VoIPController.CONNECTION_MAX_LAYER`

        Raises:
            :class:`ValueError` if either no endpoints are provided or endpoints without IPv4 or with wrong \
            ``peer_tag`` (must be either ``None`` or have length of 16 bytes) are detected
        """
        if not endpoints:
            raise ValueError('endpoints len is 0')
        for ep in endpoints:
            if ep.ip is None or not len(ep.ip):
                raise ValueError('endpoint {} has empty/null ipv4'.format(ep))
            if ep.peer_tag is not None and len(ep.peer_tag) != 16:
                raise ValueError('endpoint {} has peer_tag of wrong length'.format(ep))
        super().set_remote_endpoints(endpoints, allow_p2p, tcp, connection_max_layer)

    def get_debug_string(self) -> str:
        """
        Get debug string

            Returns:
                ``str`` containing debug info
        """
        return super().get_debug_string()

    def set_network_type(self, _type: NetType):
        """
        Set network type

        Args:
            _type (:class:`NetType`): Network type to set
        """
        super().set_network_type(_NetType(_type.value))

    def set_mic_mute(self, mute: bool):
        """
        Set "microphone" state. If muted, audio is not being sent

        Args:
            mute (``bool``): Whether to mute "microphone"
        """
        super().set_mic_mute(mute)

    def set_config(self,
                   recv_timeout: float,
                   init_timeout: float,
                   data_saving_mode: DataSaving,
                   call_id: int,
                   enable_aec: bool = True,
                   enable_ns: bool = True,
                   enable_agc: bool = True,
                   log_file_path: str = None,
                   status_dump_path: str = None,
                   log_packet_stats: bool = None):
        """
        Set call config

        Args:
            recv_timeout (``float``):
                Packet receive timeout, usually value received from ``help.getConfig()`` is used

            init_timeout (``float``):
                Packet init timeout, usually value received from ``help.getConfig()`` is used

            data_saving_mode (:class:`DataSaving`): Data saving mode

            call_id (``int``): Call ID

            enable_aec (``bool``, *optional*):
                Whether to enable automatic echo cancellation, defaults to ``True``

            enable_ns (``bool``, *optional*):
                Whether to enable noise suppression, defaults to ``True``

            enable_agc (``bool``, *optional*):
                Whether to enable automatic gain control, defaults to ``True``

            log_file_path (``str``, *optional*):
                Call log file path, calculated automatically if not provided

            status_dump_path (``str``, *optional*):
                Status dump path, calculated automatically if not provided and ``debug`` is enabled

            log_packet_stats (``bool``, *optional*):
                Whether to log packet stats, defaults to ``debug`` value
        """
        if log_file_path is None:
            if self.debug:
                log_file_path = self._get_log_file_path('voip{}'.format(call_id))
            else:
                log_file_path = self._get_log_file_path_for_call_id(call_id)  # wtf?
        if status_dump_path is None:
            status_dump_path = self._get_log_file_path('voip_stats') if self.debug else ''
        if log_packet_stats is None:
            log_packet_stats = self.debug
        super().set_config(recv_timeout, init_timeout, _DataSaving(data_saving_mode.value), enable_aec, enable_ns, enable_agc,
                           log_file_path, status_dump_path, log_packet_stats)

    def debug_ctl(self, request: int, param: int):
        """
        Debugging options

        Args:
            request (``int``):
                Option (``1`` for max bitrate, ``2`` for packet loss (in percents), ``3`` for toggling p2p, \
                ``4`` for toggling echo cancelling)

            param (``int``):
                Numeric value for options 1 and 2, ``0`` or ``1`` for options 3 and 4
        """
        super().debug_ctl(request, param)

    def get_preferred_relay_id(self) -> int:
        """
        Get preferred relay ID (used in ``discardCall`` MTProto request)

        Returns:
            ``int`` ID
        """
        return super().get_preferred_relay_id()

    def get_last_error(self) -> CallError:
        """
        Get last error type

        Returns:
            :class:`CallError` matching last occurred error type
        """
        return CallError(super().get_last_error())

    def get_stats(self) -> Stats:
        """
        Get call stats

        Returns:
            :class:`Stats` object
        """
        return super().get_stats()

    def get_debug_log(self) -> str:
        """
        Get debug log

        Returns:
            JSON ``str`` containing debug log
        """
        return super().get_debug_log()

    def set_audio_output_gain_control_enabled(self, enabled: bool):
        """
        Toggle output gain control

        Args:
            enabled (``bool``): Whether to enable output gain control
        """
        super().set_audio_output_gain_control_enabled(enabled)

    def set_echo_cancellation_strength(self, strength: int):
        """
        Set echo cancellation strength, does nothing currently but was in Java bindings (?)

        Args:
            strength (``int``): Strength value
        """
        super().set_echo_cancellation_strength(strength)

    def get_peer_capabilities(self) -> int:
        """
        Get peer capabilities

        Returns:
            ``int`` with bit mask, looks like it is used only for experimental features (group, video calls)
        """
        return super().get_peer_capabilities()

    def need_rate(self) -> bool:
        """
        Get whether the call needs rating

        Returns:
            ``bool`` value
        """
        return super().need_rate()

    @property
    def native_io(self) -> bool:
        """
        Get native I/O status (file I/O implemented in C++)

        Returns:
            ``bool`` status (enabled or not)
        """
        return self._native_io_get()

    @native_io.setter
    def native_io(self, val: bool) -> None:
        """
        Set native I/O status (file I/O implemented in C++)

        Args:
            val (``bool``): Status value
        """
        self._native_io_set(val)

    def play(self, path: str) -> bool:
        """
        Add a file to play queue for native I/O

        Args:
            path (``str``): File path

        Returns:
            ``bool`` whether opening the file was successful. File is not added to queue on failure.
        """
        return super().play(path)

    def play_on_hold(self, paths: List[str]) -> None:
        """
        Replace the hold queue for native I/O

        Args:
            paths (``list`` of ``str``): List of file paths
        """
        super().play_on_hold(paths)

    def set_output_file(self, path: str) -> bool:
        """
        Set output file for native I/O

        Args:
            path (``str``): File path

        Returns:
            ``bool`` whether opening the file was successful. Output file is not replaced on failure.
        """
        return super().set_output_file(path)

    def clear_play_queue(self) -> None:
        """
        Clear the play queue for native I/O
        """
        super().clear_play_queue()

    def clear_hold_queue(self) -> None:
        """
        Clear the hold queue for native I/O
        """
        super().clear_hold_queue()

    def unset_output_file(self) -> None:
        """
        Unset the output file for native I/O
        """
        super().unset_output_file()

    # native code callback
    def _handle_state_change(self, state: _CallState):
        state = CallState(state)

        if state == CallState.ESTABLISHED and not self.start_time:
            self.start_time = get_real_elapsed_time()

        for handler in self.call_state_changed_handlers:
            callable(handler) and handler(state)

    # native code callback
    def _handle_signal_bars_change(self, count: int):
        for handler in self.signal_bars_changed_handlers:
            callable(handler) and handler(count)

    def update_state(self, state: CallState):
        """
        Manually update state (only triggers handlers)

        Args:
            state (:class:`CallState`): State to set
        """
        self._handle_state_change(state)

    def set_send_audio_frame_callback(self, func: callable):
        """
        Set callback providing audio data to send

        Should accept one argument (``int`` length of requested audio frame) and return ``bytes`` object with audio \
        data encoded in 16-bit signed PCM

        If returned object has insufficient length, it will be automatically padded with zero bytes

        Args:
            func (``callable``): Callback function
        """
        self.send_audio_frame_callback = func

    def _send_audio_frame_impl(self, length: int):
        frame = b''
        if callable(self.send_audio_frame_callback):
            frame = self.send_audio_frame_callback(length)
        return frame.ljust(length, b'\0')

    def set_recv_audio_frame_callback(self, func: callable):
        """
        Set callback receiving incoming audio data

        Should accept one argument (``bytes``) with audio data encoded in 16-bit signed PCM

        Args:
            func (``callable``): Callback function
        """
        self.recv_audio_frame_callback = func

    def _recv_audio_frame_impl(self, frame: bytes):
        if callable(self.recv_audio_frame_callback):
            self.recv_audio_frame_callback(frame)

    def _get_log_file_path(self, name: str) -> str:
        os.makedirs(self.logs_dir, exist_ok=True)
        now = datetime.now()
        fname = '{}_{}_{}_{}_{}_{}_{}.txt'.format(now.year, now.month, now.day, now.hour, now.minute, now.second, name)
        return os.path.abspath(os.path.join(self.logs_dir, fname))

    def _get_log_file_path_for_call_id(self, call_id: int) -> str:
        os.makedirs(self.logs_dir, exist_ok=True)
        # Java version cleans up old logs (*.log) for non-debug version here (leaves 20 latest)
        return os.path.abspath(os.path.join(self.logs_dir, '{}.log'.format(call_id)))


class VoIPServerConfig(_VoIPServerConfig):
    """
    Global server config class. This class contains default config in its source
    """

    # default config
    config = {
        'audio_max_bitrate': 20000,
        'audio_max_bitrate_gprs': 8000,
        'audio_max_bitrate_edge': 16000,
        'audio_max_bitrate_saving': 8000,
        'audio_init_bitrate': 16000,
        'audio_init_bitrate_gprs': 8000,
        'audio_init_bitrate_edge': 8000,
        'audio_init_bitrate_saving': 8000,
        'audio_bitrate_step_incr': 1000,
        'audio_bitrate_step_decr': 1000,
        'audio_min_bitrate': 8000,
        'relay_switch_threshold': 0.8,
        'p2p_to_relay_switch_threshold': 0.6,
        'relay_to_p2p_switch_threshold': 0.8,
        'reconnecting_state_timeout': 2.0,
        'rate_flags': 0xFFFFFFFF,
        'rate_min_rtt': 0.6,
        'rate_min_send_loss': 0.2,
        'packet_loss_for_extra_ec': 0.02,
        'max_unsent_stream_packets': 2,
    }

    @classmethod
    def set_config(cls, _json: Union[str, dict]):
        """
        Set global server config

        Args:
            _json (``str`` | ``dict``): either JSON-encoded object or ``dict`` containing config values. \
            Might be received from MTProto ``phone.getCallConfig()`` call, if not set default values are used

        Raises:
            Prints an error to ``stderr`` if JSON parsing (for ``str`` argument) or encoding (for ``dict`` argument) \
            has occurred
        """
        try:
            if isinstance(_json, dict):
                _json = json.dumps(_json)
            cls.config.update(json.loads(_json))
            _VoIPServerConfig.set_config(_json)
        except json.JSONDecodeError as e:
            print('Error parsing VoIP config', e, file=sys.stderr)
        except TypeError as e:
            print('Error building JSON', e, file=sys.stderr)

    @classmethod
    def set_bitrate_config(cls, init_bitrate: int = 16000, max_bitrate: int = 20000, min_bitrate: int = 8000,
                           decrease_step: int = 1000, increase_step: int = 1000):
        """
        Helper method for setting bitrate options

        Args:
            init_bitrate (``int``): Initial bitrate value
            max_bitrate (``int``): Maximum bitrate value
            min_bitrate (``int``):  Minimum bitrate value
            decrease_step (``int``): Bitrate decrease step
            increase_step (``int``): Bitrate increase step

        Raises:
            Same as :meth:`set_config`
        """
        cls.set_config({
            'audio_init_bitrate': init_bitrate,
            'audio_max_bitrate': max_bitrate,
            'audio_min_bitrate': min_bitrate,
            'audio_bitrate_step_decr': decrease_step,
            'audio_bitrate_step_incr': increase_step,
        })


__all__ = ['NetType', 'DataSaving', 'CallState', 'CallError', 'Stats', 'Endpoint', 'VoIPController',
           'VoIPServerConfig']

File: src/_tgvoip.cpp
/*
 * PytgVoIP - Telegram VoIP Library for Python
 * Copyright (C) 2019 bakatrouble <https://github.com/bakatrouble>
 *
 * This file is part of PytgVoIP.
 *
 * PytgVoIP is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PytgVoIP is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with PytgVoIP.  If not, see <http://www.gnu.org/licenses/>.
 */


#include "_tgvoip.h"
#include <iostream>
#include <utility>
// #include <chrono>

Endpoint::Endpoint(int64_t id, std::string ip, std::string ipv6, uint16_t port, const std::string &peer_tag)
    : id(id), ip(std::move(ip)), ipv6(std::move(ipv6)), port(port), peer_tag(peer_tag) {}

VoIPController::VoIPController() {
    ctrl = nullptr;
    output_file = nullptr;
    native_io = false;
}

VoIPController::VoIPController(const std::string &_persistent_state_file) : VoIPController() {
    if (!_persistent_state_file.empty())
        persistent_state_file = _persistent_state_file;
}

void VoIPController::init() {
    ctrl = new tgvoip::VoIPController();
    ctrl->implData = (void *)this;
    tgvoip::VoIPController::Callbacks callbacks {};
    callbacks.connectionStateChanged = [](tgvoip::VoIPController *ctrl, int state) {
        ((VoIPController *)ctrl->implData)->_handle_state_change(CallState(state));
    };
    callbacks.signalBarCountChanged = [](tgvoip::VoIPController *ctrl, int count) {
        ((VoIPController *)ctrl->implData)->_handle_signal_bars_change(count);
    };
    callbacks.groupCallKeyReceived = nullptr;
    callbacks.groupCallKeySent = nullptr;
    callbacks.upgradeToGroupCallRequested = nullptr;
    ctrl->SetCallbacks(callbacks);
    ctrl->SetAudioDataCallbacks(
            [this](int16_t *buf, size_t size) {
                this->send_audio_frame(buf, size);
            },
            [this](int16_t *buf, size_t size) {
                this->recv_audio_frame(buf, size);
            },
            [](int16_t *buf, size_t size) {}
    );

    if (!persistent_state_file.empty()) {
        FILE *f = fopen(persistent_state_file.c_str(), "r");
        if (f) {
            fseek(f, 0, SEEK_END);
            auto len = static_cast<size_t>(ftell(f));
            fseek(f, 0, SEEK_SET);
            if(len<1024*512 && len>0){
                auto fbuf = static_cast<char *>(malloc(len));
                fread(fbuf, 1, len, f);
                std::vector<uint8_t> state(fbuf, fbuf+len);
                free(fbuf);
                ctrl->SetPersistentState(state);
            }
            fclose(f);
        }
    }
}

VoIPController::~VoIPController() {
    ctrl->Stop();
    std::vector<uint8_t> state = ctrl->GetPersistentState();
    delete ctrl;
    clear_play_queue();
    clear_hold_queue();
    unset_output_file();
    if (!persistent_state_file.empty()) {
        FILE *f = fopen(persistent_state_file.c_str(), "w");
        if (f) {
            fwrite(state.data(), 1, state.size(), f);
            fclose(f);
        }
    }
}

void VoIPController::start() {
    ctrl->Start();
}

void VoIPController::connect() {
    ctrl->Connect();
}

void VoIPController::set_proxy(const std::string &address, uint16_t port, const std::string &username, const std::string &password) {
    ctrl->SetProxy(tgvoip::PROXY_SOCKS5, address, port, username, password);
}

void VoIPController::set_encryption_key(char *key, bool is_outgoing) {
    ctrl->SetEncryptionKey(key, is_outgoing);
}

void VoIPController::set_remote_endpoints(std::list<Endpoint> endpoints, bool allow_p2p, bool tcp, int connection_max_layer) {
    std::vector<tgvoip::Endpoint> eps;
    for (auto const &ep : endpoints) {
        tgvoip::IPv4Address v4addr(ep.ip);
        tgvoip::IPv6Address v6addr("::0");
        if (!ep.ipv6.empty())
            v6addr = tgvoip::IPv6Address(ep.ipv6);
        std::string peer_tag = ep.peer_tag;
        unsigned char p_tag[16];
        if (!peer_tag.empty())
            memcpy(p_tag, peer_tag.c_str(), 16);
        eps.emplace_back(tgvoip::Endpoint(ep.id, ep.port, v4addr, v6addr,
                tcp ? tgvoip::Endpoint::Type::TCP_RELAY : tgvoip::Endpoint::Type::UDP_RELAY, p_tag));
    }
    ctrl->SetRemoteEndpoints(eps, allow_p2p, connection_max_layer);
}

std::string VoIPController::get_debug_string() {
    return ctrl->GetDebugString();
}

void VoIPController::set_network_type(NetType type) {
    ctrl->SetNetworkType(type);
}

void VoIPController::set_mic_mute(bool mute) {
    ctrl->SetMicMute(mute);
}

void VoIPController::set_config(double init_timeout, double recv_timeout, DataSaving data_saving_mode, bool enable_aec,
                                bool enable_ns, bool enable_agc,
#ifndef _WIN32
                                const std::string &log_file_path, const std::string &status_dump_path,
#else
                                const std::wstring &log_file_path, const std::wstring &status_dump_path,
#endif
                                bool log_packet_stats) {
    tgvoip::VoIPController::Config cfg;
    cfg.initTimeout = init_timeout;
    cfg.recvTimeout = recv_timeout;
    cfg.dataSaving = data_saving_mode;
    cfg.enableAEC = enable_aec;
    cfg.enableNS = enable_ns;
    cfg.enableAGC = enable_agc;
    cfg.enableCallUpgrade = false;
    if (!log_file_path.empty())
        cfg.logFilePath = log_file_path;
    if (!status_dump_path.empty())
        cfg.statsDumpFilePath = status_dump_path;
    cfg.logPacketStats = log_packet_stats;
    ctrl->SetConfig(cfg);
}

void VoIPController::debug_ctl(int request, int param) {
    ctrl->DebugCtl(request, param);
}

long VoIPController::get_preferred_relay_id() {
    return ctrl->GetPreferredRelayID();
}

CallError VoIPController::get_last_error() {
    return CallError(ctrl->GetLastError());
}

Stats VoIPController::get_stats() {
    tgvoip::VoIPController::TrafficStats _stats {};
    ctrl->GetStats(&_stats);
    return Stats {
        _stats.bytesSentWifi,
        _stats.bytesSentMobile,
        _stats.bytesRecvdWifi,
        _stats.bytesRecvdMobile,
    };
}

std::string VoIPController::get_debug_log() {
    return ctrl->GetDebugLog();
}

void VoIPController::set_audio_output_gain_control_enabled(bool enabled) {
    ctrl->SetAudioOutputGainControlEnabled(enabled);
}

void VoIPController::set_echo_cancellation_strength(int strength) {
    ctrl->SetEchoCancellationStrength(strength);
}

int VoIPController::get_peer_capabilities() {
    return ctrl->GetPeerCapabilities();
}

bool VoIPController::need_rate() {
    return ctrl->NeedRate();
}

/* std::vector<tgvoip::AudioInputDevice> VoIPController::enumerate_audio_inputs() {
    return tgvoip::VoIPController::EnumerateAudioInputs();
}

std::vector<tgvoip::AudioOutputDevice> VoIPController::enumerate_audio_outputs() {
    return tgvoip::VoIPController::EnumerateAudioOutputs();
}

void VoIPController::set_current_audio_input(std::string &id) {
    ctrl->SetCurrentAudioInput(id);
}

void VoIPController::set_current_audio_output(std::string &id) {
    ctrl->SetCurrentAudioOutput(id);
}

std::string VoIPController::get_current_audio_input_id() {
    return ctrl->GetCurrentAudioInputID();
}

std::string VoIPController::get_current_audio_output_id() {
    return ctrl->GetCurrentAudioOutputID();
} */

void VoIPController::_handle_state_change(CallState state) {
    throw py::not_implemented_error();
}

void VoIPController::_handle_signal_bars_change(int count) {
    throw py::not_implemented_error();
}

void VoIPController::send_audio_frame(int16_t *buf, size_t size) {
    tgvoip::MutexGuard m(input_mutex);
    // auto start = std::chrono::high_resolution_clock::now();
    if (native_io) {
        this->_send_audio_frame_native_impl(buf, size);
    } else {
        char *frame = this->_send_audio_frame_impl(sizeof(int16_t) * size);
        if (frame != nullptr) {
            memcpy(buf, frame, sizeof(int16_t) * size);
        }
    }
    // auto finish = std::chrono::high_resolution_clock::now();
    // std::cout << "send: " << std::chrono::duration_cast<std::chrono::nanoseconds>(finish - start).count() << std::endl;
}

char *VoIPController::_send_audio_frame_impl(ulong len) { return (char *)""; }

void VoIPController::_send_audio_frame_native_impl(int16_t *buf, size_t size) {
    if (!input_files.empty()) {
        size_t read_size = fread(buf, sizeof(int16_t), size, input_files.front());
        if (read_size != size) {
            fclose(input_files.front());
            input_files.pop();
            size_t read_offset = read_size % size;
            memset(buf + read_offset, 0, size - read_offset);
        }
    } else if (!hold_files.empty()) {
        size_t read_size = fread(buf, sizeof(int16_t), size, hold_files.front());
        if (read_size != size) {
            fseek(hold_files.front(), 0, SEEK_SET);
            hold_files.push(hold_files.front());
            hold_files.pop();
            size_t read_offset = read_size % size;
            memset(buf + read_offset, 0, size - read_offset);
        }
    }
}

void VoIPController::recv_audio_frame(int16_t *buf, size_t size) {
    tgvoip::MutexGuard m(output_mutex);
    // auto start = std::chrono::high_resolution_clock::now();
    if (buf != nullptr) {
        if (native_io) {
            this->_recv_audio_frame_native_impl(buf, size);
        } else {
            std::string frame((const char *) buf, sizeof(int16_t) * size);
            this->_recv_audio_frame_impl(frame);
        }
    }
    // auto finish = std::chrono::high_resolution_clock::now();
    // std::cout << "recv: " << std::chrono::duration_cast<std::chrono::nanoseconds>(finish - start).count() << std::endl;
}

void VoIPController::_recv_audio_frame_impl(const py::bytes &frame) {}

void VoIPController::_recv_audio_frame_native_impl(int16_t *buf, size_t size) {
    if (output_file != nullptr) {
        size_t written_size = fwrite(buf, sizeof(int16_t), size, output_file);
        if (written_size != size) {
            std::cerr << "Written size (" << written_size << ") does not match expected (" << size << ")" << std::endl;
        }
    }
}

std::string VoIPController::get_version(const py::object& /* cls */) {
    return tgvoip::VoIPController::GetVersion();
}

int VoIPController::connection_max_layer(const py::object& /* cls */) {
    return tgvoip::VoIPController::GetConnectionMaxLayer();
}

bool VoIPController::_native_io_get() {
    return native_io;
}

void VoIPController::_native_io_set(bool status) {
    native_io = status;
}

bool VoIPController::play(std::string &path) {
    FILE *tmp = fopen(path.c_str(), "rb");
    if (tmp == nullptr) {
        std::cerr << "Unable to open file " << path << " for reading" << std::endl;
        return false;
    }
    tgvoip::MutexGuard m(input_mutex);
    input_files.push(tmp);
    return true;
}

void VoIPController::play_on_hold(std::vector<std::string> &paths) {
    clear_hold_queue();
    tgvoip::MutexGuard m(input_mutex);
    for (auto &path : paths) {
        FILE *tmp = fopen(path.c_str(), "rb");
        if (tmp == nullptr) {
            std::cerr << "Unable to open file " << path << " for reading" << std::endl;
        } else {
            hold_files.push(tmp);
        }
    }
}

bool VoIPController::set_output_file(std::string &path) {
    FILE *tmp = fopen(path.c_str(), "wb");
    if (tmp == nullptr) {
        std::cerr << "Unable to open file " << path << " for writing" << std::endl;
        return false;
    }
    unset_output_file();
    tgvoip::MutexGuard m(output_mutex);
    output_file = tmp;
    return true;
}

void VoIPController::clear_play_queue() {
    tgvoip::MutexGuard m(input_mutex);
    while (!input_files.empty()) {
        fclose(input_files.front());
        input_files.pop();
    }
}

void VoIPController::clear_hold_queue() {
    tgvoip::MutexGuard m(input_mutex);
    while (!hold_files.empty()) {
        fclose(hold_files.front());
        hold_files.pop();
    }
}

void VoIPController::unset_output_file() {
    if (output_file != nullptr) {
        tgvoip::MutexGuard m(output_mutex);
        fclose(output_file);
        output_file = nullptr;
    }
}

void VoIPServerConfig::set_config(std::string &json_str) {
    tgvoip::ServerConfig::GetSharedInstance()->Update(json_str);
}
